File: joke_generator.go
// joke_generator.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

type Joke struct {
	ID       int    `json:"id"`
	Text     string `json:"text"`
	Category string `json:"category"`
}

type JokeGenerator struct {
	jokes []Joke
}

func NewJokeGenerator() *JokeGenerator {
	g := &JokeGenerator{}
	g.loadDefaultJokes()
	return g
}

func (g *JokeGenerator) loadDefaultJokes() {
	g.jokes = []Joke{
		{1, "Why don't scientists trust atoms? Because they make up everything!", "science"},
		{2, "What do you call a fake noodle? An impasta!", "food"},
		{3, "Why did the scarecrow win an award? Because he was outstanding in his field!", "pun"},
		{4, "I'm reading a book on anti-gravity. It's impossible to put down!", "book"},
		{5, "What's the best thing about Switzerland? I don't know, but the flag is a big plus!", "geography"},
		{6, "Why do programmers prefer dark mode? Because light attracts bugs!", "programming"},
		{7, "I told my wife she was drawing her eyebrows too high. She looked surprised.", "marriage"},
		{8, "What do you call a bear with no teeth? A gummy bear!", "animal"},
		{9, "Why don't eggs tell jokes? They'd crack each other up!", "food"},
		{10, "I used to play piano by ear, but now I use my hands.", "music"},
	}
}

func (g *JokeGenerator) GetRandomJoke() *Joke {
	if len(g.jokes) == 0 {
		return nil
	}
	return &g.jokes[rand.Intn(len(g.jokes))]
}

func (g *JokeGenerator) GetAllJokes() []Joke {
	return g.jokes
}

func (g *JokeGenerator) AddJoke(text, category string) Joke {
	id := 1
	if len(g.jokes) > 0 {
		id = g.jokes[len(g.jokes)-1].ID + 1
	}
	if category == "" {
		category = "general"
	}
	joke := Joke{ID: id, Text: text, Category: category}
	g.jokes = append(g.jokes, joke)
	return joke
}

func (g *JokeGenerator) RemoveJoke(id int) bool {
	for i, j := range g.jokes {
		if j.ID == id {
			g.jokes = append(g.jokes[:i], g.jokes[i+1:]...)
			return true
		}
	}
	return false
}

func (g *JokeGenerator) GetByCategory(category string) []Joke {
	var result []Joke
	for _, j := range g.jokes {
		if strings.EqualFold(j.Category, category) {
			result = append(result, j)
		}
	}
	return result
}

func (g *JokeGenerator) GetCategories() []string {
	m := make(map[string]bool)
	for _, j := range g.jokes {
		m[j.Category] = true
	}
	var cats []string
	for k := range m {
		cats = append(cats, k)
	}
	return cats
}

func (g *JokeGenerator) SaveToFile(filename string) error {
	data, err := json.MarshalIndent(g.jokes, "", "  ")
	if err != nil {
		return err
	}
	return ioutil.WriteFile(filename, data, 0644)
}

func (g *JokeGenerator) LoadFromFile(filename string) error {
	data, err := ioutil.ReadFile(filename)
	if err != nil {
		return err
	}
	return json.Unmarshal(data, &g.jokes)
}

func displayJoke(j *Joke) {
	if j == nil {
		fmt.Println("No jokes available.")
		return
	}
	fmt.Printf("\n😂 Joke: %s\n", j.Text)
	if j.Category != "" {
		fmt.Printf("📁 Category: %s\n", j.Category)
	}
}

func main() {
	rand.Seed(time.Now().UnixNano())
	gen := NewJokeGenerator()
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println("=== Random Joke Generator ===")
	for {
		fmt.Println("\n1. Get a random joke")
		fmt.Println("2. Show all jokes")
		fmt.Println("3. Add a joke")
		fmt.Println("4. Remove a joke")
		fmt.Println("5. Filter jokes by category")
		fmt.Println("6. Show categories")
		fmt.Println("7. Save jokes to file")
		fmt.Println("8. Load jokes from file")
		fmt.Println("9. Exit")
		fmt.Print("Choose: ")
		scanner.Scan()
		choice := strings.TrimSpace(scanner.Text())
		switch choice {
		case "1":
			joke := gen.GetRandomJoke()
			displayJoke(joke)
		case "2":
			jokes := gen.GetAllJokes()
			if len(jokes) == 0 {
				fmt.Println("No jokes.")
			} else {
				fmt.Println("\nAll jokes:")
				for _, j := range jokes {
					fmt.Printf("[%d] %s (%s)\n", j.ID, j.Text, j.Category)
				}
			}
		case "3":
			fmt.Print("Enter your joke: ")
			scanner.Scan()
			text := strings.TrimSpace(scanner.Text())
			if text == "" {
				fmt.Println("Joke cannot be empty.")
				continue
			}
			fmt.Print("Category (optional): ")
			scanner.Scan()
			cat := strings.TrimSpace(scanner.Text())
			joke := gen.AddJoke(text, cat)
			fmt.Printf("Joke added with ID %d.\n", joke.ID)
		case "4":
			fmt.Print("Enter joke ID to remove: ")
			scanner.Scan()
			idStr := strings.TrimSpace(scanner.Text())
			id, err := strconv.Atoi(idStr)
			if err != nil {
				fmt.Println("Invalid ID.")
				continue
			}
			if gen.RemoveJoke(id) {
				fmt.Println("Joke removed.")
			} else {
				fmt.Println("Joke not found.")
			}
		case "5":
			fmt.Print("Enter category: ")
			scanner.Scan()
			cat := strings.TrimSpace(scanner.Text())
			jokes := gen.GetByCategory(cat)
			if len(jokes) == 0 {
				fmt.Printf("No jokes in category '%s'.\n", cat)
			} else {
				fmt.Printf("\nJokes in category '%s':\n", cat)
				for _, j := range jokes {
					fmt.Printf("[%d] %s\n", j.ID, j.Text)
				}
			}
		case "6":
			cats := gen.GetCategories()
			if len(cats) == 0 {
				fmt.Println("No categories.")
			} else {
				fmt.Println("Categories:", strings.Join(cats, ", "))
			}
		case "7":
			fmt.Print("Filename: ")
			scanner.Scan()
			fname := strings.TrimSpace(scanner.Text())
			err := gen.SaveToFile(fname)
			if err != nil {
				fmt.Println("Error saving:", err)
			} else {
				fmt.Printf("Saved to %s.\n", fname)
			}
		case "8":
			fmt.Print("Filename: ")
			scanner.Scan()
			fname := strings.TrimSpace(scanner.Text())
			err := gen.LoadFromFile(fname)
			if err != nil {
				fmt.Println("Error loading:", err)
			} else {
				fmt.Printf("Loaded from %s.\n", fname)
			}
		case "9":
			fmt.Println("Goodbye! 😄")
			return
		default:
			fmt.Println("Invalid choice.")
		}
	}
}

File: joke_generator.js
// joke_generator.js
const fs = require('fs');
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

class JokeGenerator {
    constructor() {
        this.jokes = [];
        this.loadDefaultJokes();
    }

    loadDefaultJokes() {
        this.jokes = [
            { id: 1, text: "Why don't scientists trust atoms? Because they make up everything!", category: "science" },
            { id: 2, text: "What do you call a fake noodle? An impasta!", category: "food" },
            { id: 3, text: "Why did the scarecrow win an award? Because he was outstanding in his field!", category: "pun" },
            { id: 4, text: "I'm reading a book on anti-gravity. It's impossible to put down!", category: "book" },
            { id: 5, text: "What's the best thing about Switzerland? I don't know, but the flag is a big plus!", category: "geography" },
            { id: 6, text: "Why do programmers prefer dark mode? Because light attracts bugs!", category: "programming" },
            { id: 7, text: "I told my wife she was drawing her eyebrows too high. She looked surprised.", category: "marriage" },
            { id: 8, text: "What do you call a bear with no teeth? A gummy bear!", category: "animal" },
            { id: 9, text: "Why don't eggs tell jokes? They'd crack each other up!", category: "food" },
            { id: 10, text: "I used to play piano by ear, but now I use my hands.", category: "music" },
        ];
    }

    getRandomJoke() {
        if (this.jokes.length === 0) return null;
        return this.jokes[Math.floor(Math.random() * this.jokes.length)];
    }

    getAllJokes() { return this.jokes; }

    addJoke(text, category = "general") {
        const id = this.jokes.length > 0 ? this.jokes[this.jokes.length-1].id + 1 : 1;
        const joke = { id, text, category };
        this.jokes.push(joke);
        return joke;
    }

    removeJoke(id) {
        const idx = this.jokes.findIndex(j => j.id === id);
        if (idx !== -1) {
            this.jokes.splice(idx, 1);
            return true;
        }
        return false;
    }

    getByCategory(category) {
        return this.jokes.filter(j => j.category.toLowerCase() === category.toLowerCase());
    }

    getCategories() {
        const cats = new Set(this.jokes.map(j => j.category));
        return [...cats].sort();
    }

    saveToFile(filename) {
        fs.writeFileSync(filename, JSON.stringify(this.jokes, null, 2), 'utf8');
    }

    loadFromFile(filename) {
        try {
            const data = fs.readFileSync(filename, 'utf8');
            this.jokes = JSON.parse(data);
        } catch (e) {
            console.log(`Error loading ${filename}: ${e.message}`);
        }
    }
}

function ask(question) {
    return new Promise(resolve => rl.question(question, resolve));
}

function displayJoke(joke) {
    if (!joke) {
        console.log("No jokes available.");
        return;
    }
    console.log(`\n😂 Joke: ${joke.text}`);
    if (joke.category) console.log(`📁 Category: ${joke.category}`);
}

async function main() {
    const gen = new JokeGenerator();
    console.log("=== Random Joke Generator ===");
    while (true) {
        console.log("\n1. Get a random joke");
        console.log("2. Show all jokes");
        console.log("3. Add a joke");
        console.log("4. Remove a joke");
        console.log("5. Filter jokes by category");
        console.log("6. Show categories");
        console.log("7. Save jokes to file");
        console.log("8. Load jokes from file");
        console.log("9. Exit");
        const choice = await ask("Choose: ");
        switch (choice.trim()) {
            case "1": {
                const joke = gen.getRandomJoke();
                displayJoke(joke);
                break;
            }
            case "2": {
                const jokes = gen.getAllJokes();
                if (jokes.length === 0) console.log("No jokes.");
                else {
                    console.log("\nAll jokes:");
                    jokes.forEach(j => console.log(`[${j.id}] ${j.text} (${j.category})`));
                }
                break;
            }
            case "3": {
                const text = await ask("Enter your joke: ");
                if (!text.trim()) { console.log("Joke cannot be empty."); break; }
                const cat = await ask("Category (optional): ");
                const joke = gen.addJoke(text.trim(), cat.trim() || "general");
                console.log(`Joke added with ID ${joke.id}.`);
                break;
            }
            case "4": {
                const idStr = await ask("Enter joke ID to remove: ");
                const id = parseInt(idStr, 10);
                if (isNaN(id)) { console.log("Invalid ID."); break; }
                if (gen.removeJoke(id)) console.log("Joke removed.");
                else console.log("Joke not found.");
                break;
            }
            case "5": {
                const cat = await ask("Enter category: ");
                const jokes = gen.getByCategory(cat);
                if (jokes.length === 0) console.log(`No jokes in category '${cat}'.`);
                else {
                    console.log(`\nJokes in category '${cat}':`);
                    jokes.forEach(j => console.log(`[${j.id}] ${j.text}`));
                }
                break;
            }
            case "6": {
                const cats = gen.getCategories();
                if (cats.length === 0) console.log("No categories.");
                else console.log("Categories:", cats.join(", "));
                break;
            }
            case "7": {
                const fname = await ask("Filename: ");
                gen.saveToFile(fname);
                console.log(`Saved to ${fname}.`);
                break;
            }
            case "8": {
                const fname = await ask("Filename: ");
                gen.loadFromFile(fname);
                console.log(`Loaded from ${fname}.`);
                break;
            }
            case "9":
                console.log("Goodbye! 😄");
                rl.close();
                return;
            default:
                console.log("Invalid choice.");
        }
    }
}

main().catch(console.error);
